File: book/DFS-BFS/p152.py
from collections import deque

def solution(all_path:str):
    answer = 1
    
    # 경로 graph화
    path = all_path.split('\n')
    graph = [[int(g) for g in p] for p in path]
    
    # 종착지
    n = len(graph)
    m = len(graph[0])
    
    # 출발지
    graph[0][0] = 0
    queue = deque([(0,0)])
    
    while queue:
        q = queue.popleft()
        
        is_append = False
        y = q[0]
        x = q[1]
        
        if y == n-1 and x == m-1:
            return answer + 1
        
        if x+1 < m:
            if y == n-1 and x+1 == m-1:
                return answer + 1
            
            if graph[y][x+1] == 1:
                graph[y][x+1] = 0
                queue.append((y, x+1))
                is_append = True
                
        if x-1 >= 0:
            if y == n-1 and x-1 == m-1:
                return answer + 1
            
            if graph[y][x-1] == 1:
                graph[y][x-1] = 0
                queue.append((y, x+1))
                is_append = True
                
        if y+1 < n:
            if y+1 == n-1 and x == m-1:
                return answer + 1
            
            if graph[y+1][x] == 1:
                graph[y+1][x] = 0
                queue.append((y+1, x))
                is_append = True
                
        if y-1 >= 0:
            if y-1 == n-1 and x == m-1:
                return answer + 1
            
            if graph[y-1][x] == 1:
                graph[y-1][x] = 0
                queue.append((y-1, x))
                is_append = True
                
        if is_append:
            # 여러개가 추가되어도 한번 이동으로 간주
            answer += 1
        
path = '''101010
111111
000001
111111
111111'''

print(solution(path))

File: book/java/src/y24/dfs_bfs/excercise/P152.java
package y24.dfs_bfs.excercise;

import java.util.ArrayDeque;
import java.util.Deque;

public class P152 {
    private static int[][] movements = {
            {1, 0},
            {-1, 0},
            {0, 1},
            {0, -1}
    };
    private static int[][] map = {
            {1, 1, 0, 0, 1, 0},
            {0, 1, 0, 1, 1, 1},
            {1, 1, 0, 0, 1, 0},
            {1, 0, 0, 1, 1, 1},
            {1, 1, 1, 1, 0, 1}
    };

    public static void main(String[] args) {
        int result = solution(new int[]{0, 0}, 5, 6);
        System.out.println("result = " + result);
    }

    private static int solution(int[] start, int n, int m) {
        int answer = 0;

        Deque<int[]> queue = new ArrayDeque();
        queue.addLast(start);

        map[start[0]][start[1]] = 0;

        int[] lastNode = start;

        while (!queue.isEmpty()) {
            int[] path = queue.pop();

            if (path[0] == n-1 && path[1] == m-1) {
                return ++answer;
            }

            for (int[] move : movements) {
                int next_i = path[0] + move[0];
                int next_j = path[1] + move[1];
                if (next_i < 0 || next_i >= n || next_j < 0 || next_j >= m) {
                    continue;
                }
                if (map[next_i][next_j] == 1) {
                    queue.addLast(new int[]{next_i, next_j});
                    map[next_i][next_j] = 0;
                }
            }
            if (!queue.contains(lastNode)) {
                lastNode = queue.getLast();
                answer += 1;
            }
        }
        return 0;
    }
}
